npm/pkg/dataplane/dpshim/dpshim.go

```diff
@@ -1,77 +1,590 @@
+	npmerrors "github.com/Azure/azure-container-networking/npm/util/errors"
+	"k8s.io/klog"
 )
 
-// TODO setting this up to unblock another workitem
-type DPShim struct {
-	outChannel chan *protos.Events
-}
+const cleanEmptySetsInHrs = 24
 
 var ErrChannelUnset = errors.New("channel must be set")
 
+type DPShim struct {
+	outChannel  chan *protos.Events
+	setCache    map[string]*controlplane.ControllerIPSets
+	policyCache map[string]*policies.NPMNetworkPolicy
+	dirtyCache  *dirtyCache
+	sync.Mutex
+}
+
 func NewDPSim(outChannel chan *protos.Events) (*DPShim, error) {
 	if outChannel == nil {
 		return nil, fmt.Errorf("out channel must be set: %w", ErrChannelUnset)
 	}
-	return &DPShim{outChannel: outChannel}, nil
-}
 
-func (dp *DPShim) InitializeDataPlane() error {
-	return nil
+	return &DPShim{
+		outChannel:  outChannel,
+		setCache:    make(map[string]*controlplane.ControllerIPSets),
+		policyCache: make(map[string]*policies.NPMNetworkPolicy),
+		dirtyCache:  newDirtyCache(),
+	}, nil
 }
 
 func (dp *DPShim) ResetDataPlane() error {
 	return nil
 }
 
-func (dp *DPShim) RunPeriodicTasks() {}
+// HydrateClients is used in DPShim to hydrate a restarted Daemon Client
+func (dp *DPShim) HydrateClients() (*protos.Events, error) {
+	dp.Lock()
+	defer dp.Unlock()
+
+	if len(dp.setCache) == 0 && len(dp.policyCache) == 0 {
+		klog.Infof("HydrateClients: No local cache objects to hydrate daemon client")
+		return nil, nil
+	}
+
+	goalStates := make(map[string]*protos.GoalState)
+
+	toApplySets, err := dp.hydrateSetCache()
+	if err != nil {
+		return nil, err
+	}
+	if toApplySets != nil {
+		goalStates[controlplane.IpsetApply] = toApplySets
+	}
+
+	toApplyPolicies, err := dp.hydratePolicyCache()
+	if err != nil {
+		return nil, err
+	}
+	if toApplyPolicies != nil {
+		goalStates[controlplane.PolicyApply] = toApplyPolicies
+	}
+
+	if len(goalStates) == 0 {
+		klog.Info("HydrateClients: No changes to apply")
+		return nil, nil
+	}
+
+	return &protos.Events{
+		Payload: goalStates,
+	}, nil
+}
+
+func (dp *DPShim) RunPeriodicTasks() {
+	// Here Run periodic task to check if any sets with empty references are present and delete them
+}
 
 func (dp *DPShim) GetIPSet(setName string) *ipsets.IPSet {
 	return nil
 }
 
+func (dp *DPShim) getIPSet(setName string) *controlplane.ControllerIPSets {
+	return dp.setCache[setName]
+}
+
+func (dp *DPShim) setExists(setName string) bool {
+	_, ok := dp.setCache[setName]
+	return ok
+}
+
 func (dp *DPShim) CreateIPSets(setNames []*ipsets.IPSetMetadata) {
+	dp.Lock()
+	defer dp.Unlock()
+	for _, set := range setNames {
+		dp.createIPSet(set)
+	}
+}
+
+func (dp *DPShim) createIPSet(set *ipsets.IPSetMetadata) {
+	setName := set.GetPrefixName()
+
+	if dp.setExists(setName) {
+		return
+	}
+
+	dp.setCache[setName] = controlplane.NewControllerIPSets(set)
+	dp.dirtyCache.modifyAddorUpdateSets(setName)
 }
 
 func (dp *DPShim) DeleteIPSet(setMetadata *ipsets.IPSetMetadata) {
+	dp.Lock()
+	defer dp.Unlock()
```

Comment:
might make more sense to move the lock/unlock to the helper functions `deleteIPSet`

Reply:
I am following the practice of Exposed function deal with locks and internal functions to not worry about locks. Internal functions by default we can assume that they are to be handled with care. Also, createIPSet is called from some external functions and we cannot deal with locking and unlocking always. Wdyt ?